BJS2-77030 конвертер валют

VladimirEkgardt opened the pull request.

VladimirEkgardt committed:
BJS2-77030 конвертер валют

SnurnicynKonstantin commented on src/main/java/faang/school/paymentservice/service/PaymentService.java:
```diff
@@ -0,0 +1,55 @@
+@RequiredArgsConstructor
+@Slf4j
+public class PaymentService {
+    private static final String APP_ID = "0134aec31a134dec86124871d5fc0cbf";
```
Это в проперти надо поместить

SnurnicynKonstantin commented on src/main/java/faang/school/paymentservice/service/PaymentService.java:
```diff
@@ -0,0 +1,55 @@
+
+    private static JSONObject getJsonObject(String urlStr) throws IOException {
+        URL url = new URL(urlStr);
+        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
```
А почему FeignClient не сиспользовал?
Что-то типа этого:
```
@FeignClient(name = "currency-converter", url = "${currency.exchange.url}")
public interface ConverterClient {
    @GetMapping(value = "api/latest.json", produces = "application/json", consumes = "application/json")
    CurrencyExchangeResponse getCurrentCurrencyExchangeRate(@RequestParam(value = "app_id") String appId);
}
```
И зависимости не надо будет тащить + кода меньше будет

SnurnicynKonstantin commented on src/test/java/faang/school/paymentservice/service/PaymentServiceTest.java:
```diff
@@ -0,0 +1,25 @@
+        PaymentRequest dto = new PaymentRequest(1, new BigDecimal("5"), Currency.USD);
+        BigDecimal rub = new BigDecimal("397.94");
+
+        BigDecimal result = service.convertToRUB(dto);
```
А здесь реальный запрос уходит в интернет? Если так, то так не следует делать. Юнит-тесты - это изолированные тесты. Работу клиента надо мокать.

VladimirEkgardt committed:
1) Переменные вынес в properties
2) Изменил обращение к сервису на FeignClient
2.1) Добавил interface ConverterClient
2.2) Добавил class ConverterConfig
2.3) Переписал PaymentService под новую реализацию
3) Скорректировал тесты

SnurnicynKonstantin commented on src/main/java/faang/school/paymentservice/service/PaymentService.java:
```diff
@@ -0,0 +1,32 @@
+
+    public BigDecimal convertToRUB(PaymentRequest dto) {
+        CurrencyExchangeResponse response =
+                converterClient.getCurrentCurrencyExchangeRate(appId, dto.currency().name(), "RUB");
```
"RUB" в константу надо перенести

SnurnicynKonstantin commented on src/main/java/faang/school/paymentservice/service/PaymentService.java:
```diff
@@ -0,0 +1,32 @@
+                converterClient.getCurrentCurrencyExchangeRate(appId, dto.currency().name(), "RUB");
+
+        BigDecimal rub = BigDecimal.valueOf(response.getRates().get("RUB"));
+        BigDecimal percent = new BigDecimal("1.01");
```
В константу - "1.01"

SnurnicynKonstantin commented on src/main/java/faang/school/paymentservice/entity/CurrencyExchangeResponse.java:
```diff
@@ -0,0 +1,13 @@
+
+@Data
+@AllArgsConstructor
+public class CurrencyExchangeResponse {
```
Эту дто надо перенести в dto пакет

SnurnicynKonstantin commented on src/main/java/faang/school/paymentservice/config/ConverterClient.java:
```diff
@@ -0,0 +1,15 @@
+
+@FeignClient(name = "currency-converter", url = "${url.service}")
+public interface ConverterClient {
+    @GetMapping(value = "api/latest.json", produces = "application/json", consumes = "application/json")
```
То, что переделал на FeignClient это очень круто!

VladimirEkgardt committed:
1) в PaymentService вынес переменные с валютой RUB и 1.01 в yaml
2) Перенес CurrencyExchangeResponse в пакет с дто

SnurnicynKonstantin commented on src/main/java/faang/school/paymentservice/service/PaymentService.java:
```diff
@@ -0,0 +1,41 @@
+
+    public BigDecimal convertToRUB(PaymentRequest dto) {
+        CurrencyExchangeResponse response =
+                converterClient.getCurrentCurrencyExchangeRate(appId, dto.currency().name(), "RUB");
```
Здесь то же можно currency использовать вместо "RUB"

SnurnicynKonstantin commented on src/main/java/faang/school/paymentservice/config/ConverterConfig.java:
```diff
@@ -0,0 +1,11 @@
+import org.springframework.context.annotation.Bean;
+import org.springframework.http.client.support.BasicAuthenticationInterceptor;
+
+public class ConverterConfig {
```
А нужен ли этот конфиг? И без него все прекрасно работает

VladimirEkgardt replied:
А я правильно понимаю, что такой конфиг и в целом бин используются для аутентификации в сторонних сервисах?

SnurnicynKonstantin commented on build.gradle.kts:
```diff
@@ -37,6 +37,9 @@ dependencies {
     testImplementation("org.springframework.boot:spring-boot-starter-test")
+
+    implementation ("com.squareup.okhttp3:okhttp:4.9.3")
+    implementation("org.json:json:20250517")
```
Похоже теперь можно удалить две эти зависимости

VladimirEkgardt committed:
1) Удалил не нужные зависимости в gradle
2) В PaymentService в convertToRUB поменял "RUB" на переменную currency
3) Удалил ConverterConfig

SnurnicynKonstantin reviewed (approved):
Классную задачку сделал с интеграцией
